### PR TITLE
8303684: Lift upcall sharing mechanism to AbstractLinker (mainline)

### PR DESCRIPTION
**Pull request by JornVernee**
Port of: https://github.com/openjdk/panama-foreign/pull/791 which lifts the sharing mechanism for upcall stubs to AbstractLinker.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303684](https://bugs.openjdk.org/browse/JDK-8303684): Lift upcall sharing mechanism to AbstractLinker (mainline)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [0355ab19](https://git.openjdk.org/jdk/pull/12883/files/0355ab19ffe8c7960b3e0187f63a19badc788cd8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12883/head:pull/12883` \
`$ git checkout pull/12883`

Update a local copy of the PR: \
`$ git checkout pull/12883` \
`$ git pull https://git.openjdk.org/jdk pull/12883/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12883`

View PR using the GUI difftool: \
`$ git pr show -t 12883`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12883.diff">https://git.openjdk.org/jdk/pull/12883.diff</a>

</details>

### CODE REVIEW

**Review by mcimadamore — APPROVED** [review]
Looks good (already reviewed on Panama repo)

**Comment by JornVernee** [comment]
@feilongjiang since this contains changes to the RISCV code as well, could you test the patch? Thanks.

**Comment by feilongjiang** [comment]
I'll submit a test on RISC-V board to see the results. It may take a while.

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64Linker.java [hunk 1]
```diff
@@ -58,8 +57,8 @@ protected MethodHandle arrangeDowncall(MethodType inferredMethodType, FunctionDe
     }
 
     @Override
-    protected MemorySegment arrangeUpcall(MethodHandle target, MethodType targetType, FunctionDescriptor function, SegmentScope scope) {
-        return  CallArranger.WINDOWS.arrangeUpcall(target, targetType, function, scope);
+    protected UpcallStubFactory arrangeUpcall(MethodType targetType, FunctionDescriptor function) {
+        return  CallArranger.WINDOWS.arrangeUpcall(targetType, function);
```
Comment (turbanoff) [inline src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64Linker.java:61]:
```suggestion
        return CallArranger.WINDOWS.arrangeUpcall(targetType, function);
```

**Pull request closed without merging**